File: plugin/zeppelin/src/main/scala/org/intellij/plugin/zeppelin/idea/wizard/ZeppelinModuleBuilder.scala
package org.intellij.plugin.zeppelin.idea.wizard

import com.intellij.ide.util.projectWizard._
import com.intellij.openapi.module.ModuleType
import com.intellij.openapi.roots.ModifiableRootModel
import com.intellij.openapi.roots.ui.configuration.ModulesProvider
import com.intellij.openapi.startup.StartupManager
import com.intellij.openapi.vfs.LocalFileSystem
import org.intellij.plugin.zeppelin.extensionpoints.TemplateFileCreator

class ZeppelinModuleBuilder extends JavaModuleBuilder {
  override def createWizardSteps(wizardContext: WizardContext,
                                 modulesProvider: ModulesProvider): Array[ModuleWizardStep] = {
    ModuleWizardStep.EMPTY_ARRAY
  }

  override def getModuleType: ModuleType[_ <: ModuleBuilder] = ZeppelinModuleType.getModuleType

  import com.intellij.ide.util.projectWizard.{ModuleWizardStep, SettingsStep}
  import com.intellij.openapi.module.StdModuleTypes

  override def modifyProjectTypeStep(settingsStep: SettingsStep): ModuleWizardStep = {
    StdModuleTypes.JAVA.modifyProjectTypeStep(settingsStep, this)
  }

  override def modifySettingsStep(settingsStep: SettingsStep): ModuleWizardStep = {
    new ZeppelinModuleSettingStep(this, settingsStep)
  }

  override def setupRootModel(rootModel: ModifiableRootModel): Unit = {
    super
      .setupRootModel(rootModel)
    val contentEntryPath = getContentEntryPath
    if (contentEntryPath == null) return

    val srcPath = contentEntryPath + "/src"
    val contentRoot = LocalFileSystem.getInstance().findFileByPath(srcPath)
    if (contentRoot == null) return


    val module = rootModel.getModule
    val project = module.getProject
    StartupManager.getInstance(project).runWhenProjectIsInitialized(() => {
      TemplateFileCreator.getAll.foreach(it => it.create(srcPath, project))
    })
  }
}

File: zeppelin/src/org/intellij/plugin/zeppelin/idea/wizard/ZeppelinModuleBuilder.kt
package org.intellij.plugin.zeppelin.idea.wizard

import com.intellij.ide.util.projectWizard.JavaModuleBuilder
import com.intellij.ide.util.projectWizard.ModuleWizardStep
import com.intellij.ide.util.projectWizard.SettingsStep
import com.intellij.ide.util.projectWizard.WizardContext
import com.intellij.openapi.module.Module
import com.intellij.openapi.module.ModuleType
import com.intellij.openapi.module.StdModuleTypes
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ModifiableRootModel
import com.intellij.openapi.roots.ui.configuration.ModulesProvider
import com.intellij.openapi.startup.StartupManager
import org.intellij.plugin.zeppelin.extensionpoints.TemplateFileCreator

open class ZeppelinModuleBuilder : JavaModuleBuilder() {
    override fun createWizardSteps(wizardContext: WizardContext,
                                   modulesProvider: ModulesProvider): Array<out ModuleWizardStep>? {
        return ModuleWizardStep.EMPTY_ARRAY
    }

    override fun getModuleType(): ModuleType<*> = ZeppelinModuleType.getModuleType()
    override fun modifyProjectTypeStep(settingsStep: SettingsStep): ModuleWizardStep {
        return StdModuleTypes.JAVA.modifyProjectTypeStep(settingsStep, this)!!
    }

    override fun modifySettingsStep(settingsStep: SettingsStep): ModuleWizardStep {
        return ZeppelinModuleSettingStep(this, settingsStep)
    }

    override fun setupRootModel(rootModel: ModifiableRootModel) {
        super.setupRootModel(rootModel)
        val contentEntryPath: String = this.contentEntryPath ?: return
        val srcPath = "$contentEntryPath/src"
        val module: Module = rootModel.module
        val project: Project = module.project
        StartupManager.getInstance(project)
            .runWhenProjectIsInitialized { TemplateFileCreator.get()?.create(srcPath, project) }
    }
}